sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs

```diff
@@ -0,0 +1,419 @@
+                    return false;
+                }
+            }
+
+        }
+
+        private async Task<bool> ImdsAvailableAsync(CancellationToken cancellationToken)
+        {
+            // send a request without the Metadata header.  This will result in a failed request,
+            // but we're just interested in if we get a response before the timeout of 500ms
+            // if we don't get a response we assume the imds endpoint is not available
+            using (Request request = _pipeline.CreateRequest())
+            {
+                request.Method = HttpPipelineMethod.Get;
+
+                request.UriBuilder.Uri = ImdsEndpoint;
+
+                request.UriBuilder.AppendQuery("api-version", ImdsApiVersion);
+
+                var imdsTimeout = new CancellationTokenSource(500).Token;
+
+                try
+                {
+                    var response = await _pipeline.SendRequestAsync(request, CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, imdsTimeout).Token).ConfigureAwait(false);
+
+                    return true;
+                }
+                // we only want to handle the case when the imdsTimeout resulted in the request being cancelled.
+                // this indicates that the request timed out and that imds is not available.  Otherwise the operation
+                // was user cancelled.  In this case we don't wan't to handle the exception so s_identityAvailable will
+                // remain unset, as we still haven't determined if the imds endpoint is available.
+                catch (OperationCanceledException ex) when (ex.CancellationToken == imdsTimeout)
+                {
+                    return false;
+                }
+            }
+        }
+
+        private Request CreateImdsAuthRequest(string[] scopes, string clientId)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            Request request = _pipeline.CreateRequest();
+
+            request.Method = HttpPipelineMethod.Get;
+
+            request.Headers.Add("Metadata", "true");
+
+            request.UriBuilder.Uri = s_endpoint;
+
+            request.UriBuilder.AppendQuery("api-version", ImdsApiVersion);
+
+            request.UriBuilder.AppendQuery("resource", Uri.EscapeDataString(resource));
+
+            if (!string.IsNullOrEmpty(clientId))
+            {
+                request.UriBuilder.AppendQuery("client_id", Uri.EscapeDataString(clientId));
+            }
+
+            return request;
+        }
+
+        private Request CreateAppServiceAuthRequest(string[] scopes, string clientId)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            Request request = _pipeline.CreateRequest();
+
+            request.Method = HttpPipelineMethod.Get;
+
+            request.Headers.Add("secret", s_secret);
```

Comment:
Docs claim the secret rotates. I don't know whether that requires an application restart so I get the value from the environment for each request.

Reply:
good catch. updated